### PR TITLE
TINY-4549: Added the ability to search and replace within a selection

### PR DESCRIPTION
**Pull request by lnewson**
This adds the new search and replace in selection feature, while also rewriting how matches are found and marked. I did originally spike out a version using `DomSearch` from pheonix, but it had a number of issues such as:
- Bundle size was massive (increased size by 180%, compared to 70% with this PR). Current is ~10-11kb minified.
- Didn't allow the level of filtering/matching needed to be comparable to the old version (eg the `isHidden` logic in `TextCollect`).
- Didn't allow searching via ranges.
- It split the text nodes while also finding the matches which messed with selection.
- Did a lot of additional looping through text elements, that wasn't needed for this use case.
As such I took the concepts from that and wrote this using both the concepts and TinyMCE APIs where possible.

Known caveats atm:
- Not the best experience on IE 11, as being single selection you cannot see the selection when the dialog opens. The find still works however, but we may want to consider showing a fake selection as a followup improvement (as discussed on slack).
- <strike>Doesn't work with "Whole words" if the selection is in the middle of a word (eg `|^^ ^^ ^^|^^` would find 3 matches for `^^` instead of 2). I know how to solve this, but this is a fair edge case and it's somewhat difficult to get right.</strike>

Note: Marking it at draft, as I'm still testing a couple more things but I wanted to get eyes on this.

### CODE REVIEW

**Comment by lnewson** [comment]
Fixed the second caveat above that I wanted to fix, so marking this as ready for review

File: modules/tinymce/src/plugins/searchreplace/main/ts/core/TextPosition.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+const extract = (elements: Element<DomText>[], matches: Position[]): TextMatch[][] => {
+  // Walk over each text node and compare with the matches
+  const nodePositions = Arr.foldl(elements, (acc, element) => {
+    const content = Text.get(element);
+    const start = acc.last;
+    const finish = start + content.length;
+
+    // Find positions for any matches in the current text node
+    const positions = Arr.bind(matches, (matches, matchIdx) => {
```
Comment (ltrouton) [inline modules/tinymce/src/plugins/searchreplace/main/ts/core/TextPosition.ts:49]: Would `match` be a better name for the first argument of the callback?

Comment (lnewson) [reply modules/tinymce/src/plugins/searchreplace/main/ts/core/TextPosition.ts:49]: Oops, yes. Thanks for catching that!

Comment (lnewson) [reply modules/tinymce/src/plugins/searchreplace/main/ts/core/TextPosition.ts]: Fixed :+1:

**Pull request merged**